### PR TITLE
feat: data augmentation for training segmentation models with torch backend

### PR DESCRIPTION
**Pull request by beniz**
This PR brings:
- basic data augmentation (rotation, cutout, perspectives) to segmentation with torch
- swapped jpg compression (quality 100) to png with compression 0, as jpg was corrupting the segmentation masks (even with quality set to 100...)

seg_aug_geom1_src
seg_aug_geom1

### CODE REVIEW

**Comment by Bycob** [comment]
PNG compression is always lossless, the compression level sets the trade-off between image size and compression time
A compression level of 0 means the images will be poorly (not?) compressed but very fast. This could lead to huge databases, which is something we want to avoid. I suggest to keep the default compression level of OpenCV (I think it's 1)

And I think jpg compression may still be a good feature to have for big datasets, maybe we can add a `lossless` boolean to `image_to_stringstream` and set it to true only for segmentation targets?

**Comment by beniz** [comment]
Yes, absolutely, but masks must not be compressed.

**Comment by Bycob** [comment]
PNG lossless compression does not alter the image in any way, so compressing masks in PNG at any level is fine, unless there is something I missed?

**Comment by beniz** [comment]
@Bycob see last commit. The most time spent is decoding the images every epoch, the encoding is done once when building the db.

File: src/backends/torch/torchdataset.cc [hunk 1]
```diff
@@ -119,11 +119,23 @@ namespace dd
   }
 
   void TorchDataset::image_to_stringstream(const cv::Mat &img,
-                                           std::ostringstream &dstream)
+                                           std::ostringstream &dstream,
+                                           const bool &compression)
```
Comment (Bycob) [inline src/backends/torch/torchdataset.cc:123]: I'd call it "lossless", in both cases the image is compressed, but with jpeg the image can be altered, while with png the image remains exactly the same

Comment (beniz) [reply src/backends/torch/torchdataset.cc:123]: In practice PNG is an order of magnitude bigger than JPG, very impacting on datasets with millions of images. Additionnally it doesn't matter the image is compressed for training, since the neural net much often would need to accomodate that type of distorsion.
Ground truth, aka masks however must not be compressed.

Comment (Bycob) [reply src/backends/torch/torchdataset.cc:123]: I do agree with what you say, that's why I asked for keeping images compressed as jpg.
However there is a vocabulary misunderstanding here:
masks must not be *altered*, however we can *compress* them.
PNG *is* a compressed format, as opposed to bitmap, but it's a *lossless* one, which means it reduces the size of the image without altering it at all. In another hand, jpg produces much smaller images but alters the image, so we can not use it for masks.

In both cases the image *is* compressed, so the parameter name `compression` is misleading, it would be preferable to rename it `lossless`.

Comment (beniz) [reply src/backends/torch/torchdataset.cc:123]: OK!

File: src/backends/torch/torchdataset.cc [hunk 2]
```diff
@@ -119,11 +119,23 @@ namespace dd
...
-                                           std::ostringstream &dstream)
+                                           std::ostringstream &dstream,
+                                           const bool &lossless)
   {
     std::vector<uint8_t> buffer;
-    std::vector<int> param = { cv::IMWRITE_JPEG_QUALITY, 100 };
-    cv::imencode(".jpg", img, buffer, param);
+    std::vector<int> param;
+    std::string ext;
+    if (compression)
```
Comment (Bycob) [inline src/backends/torch/torchdataset.cc:128]: don't forget to change it here :)
and lossless == png so the branches need to be exchanged

**Review by fantes — APPROVED** [review]

File: src/backends/torch/torchdataaug.h [hunk 3]
```diff
@@ -151,7 +151,11 @@ namespace dd
     float _prob = 0.0;
     bool _geometry_persp_horizontal
         = true; /**< horizontal perspective change. */
+    // bool _geometry_persp_horizontal_right = true; /**< right or left, when
```
Comment (Bycob) [inline src/backends/torch/torchdataaug.h:154]: why these are commented out?

Comment (beniz) [reply src/backends/torch/torchdataaug.h:154]: I will remove them, unneeded.

File: src/backends/torch/torchdataaug.h [hunk 4]
```diff
@@ -161,11 +161,15 @@ namespace dd
...
-                                       mirrored, 3: repeat nearest. */
+                                       repeat nearest (replicate). */
     float _geometry_bbox_intersect
         = 0.75; /**< warped bboxes must at least have a 75% intersect with the
                    original bbox, otherwise they are filtered out.*/
     cv::Mat _lambda; /**< warp perspective matrix. */
+    // float quads_x0 = -1.0; /**< quad on x axis, stored for target. */
+    // float quads_y0 = -1.0; /**< quad on y axis, stored for target. */
+    // cv::Point2f _inputQuad[4];
+    // cv::Point2f _outputQuad[4];
```
Comment (Bycob) [inline src/backends/torch/torchdataaug.h:172]: Remove this too?

Comment (beniz) [reply src/backends/torch/torchdataaug.h]: done

**Review by Bycob — APPROVED** [review]
